src/Microsoft.IdentityModel.Tokens/Encryption/RsaKeyWrapProvider.cs

```diff
@@ -0,0 +1,459 @@
+// THE SOFTWARE.
+//
+//------------------------------------------------------------------------------
+
+using System;
+using System.Security.Cryptography;
+using System.Globalization;
+using Microsoft.IdentityModel.Logging;
+
+namespace Microsoft.IdentityModel.Tokens
+{
+    /// <summary>
+    /// Provides RSA Wrap key and Unwrap key services.
+    /// </summary>
+    public class RsaKeyWrapProvider : IDisposable
+    {
+#if NETSTANDARD1_4
+        private RSA _rsa;
+#else
+        private RSACryptoServiceProvider _rsaCryptoServiceProvider;
+#endif
+        private bool _disposeRsa;
+        private RSACryptoServiceProviderProxy _rsaCryptoServiceProviderProxy;
+        private bool _disposed = false;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="RsaKeyWrapProvider"/> class used for wrap key and unwrap key.
+        /// <param name="key">The <see cref="SecurityKey"/> that will be used for crypto operations.</param>
+        /// <param name="algorithm">The KeyWrap algorithm to apply.</param>
+        /// <param name="willDecrypt">Whether this <see cref="RsaKeyWrapProvider"/> is required to create decrypts then set this to true.</param>
+        /// <exception cref="ArgumentNullException">'key' is null.</exception>
+        /// <exception cref="ArgumentNullException">'algorithm' is null.</exception>
+        /// <exception cref="ArgumentOutOfRangeException">The keysize doesn't match the algorithm.</exception>
+        /// <exception cref="ArgumentOutOfRangeException">The <see cref="SecurityKey"/>  is not supported.</exception>
+        /// <exception cref="ArgumentException">If <see cref="SecurityKey"/> and algorithm pair are not supported.</exception>
+        /// <exception cref="InvalidOperationException">Failed to create RSA algorithm with provided key and algorithm.</exception>
+        /// </summary>
+        public RsaKeyWrapProvider(SecurityKey key, string algorithm, bool willDecrypt)
+        {
+            if (key == null)
+                throw LogHelper.LogArgumentNullException(nameof(key));
+
+            if (string.IsNullOrEmpty(algorithm))
+                throw LogHelper.LogArgumentNullException(nameof(algorithm));
+
+            if (key.KeySize < 2048)
+            {
+                string keyId = key.KeyId ?? string.Empty;
+                throw LogHelper.LogExceptionMessage(new ArgumentOutOfRangeException("key.KeySize", string.Format(CultureInfo.InvariantCulture, LogMessages.IDX10662, algorithm, 2048, keyId, key.KeySize)));
+            }
+
+            if (!IsSupportedAlgorithm(key, algorithm))
+                throw LogHelper.LogExceptionMessage(new ArgumentException(string.Format(CultureInfo.InvariantCulture, LogMessages.IDX10671, algorithm, key)));
+
+            Algorithm = algorithm;
+            Key = key;
+
+            ResolveRsaAlgorithm(key, algorithm, willDecrypt);
+#if NETSTANDARD1_4
+            if (_rsa == null && _rsaCryptoServiceProviderProxy == null)
```

Comment:
_rsaCryptoServiceProviderProxy should only be in scope fro NET451. See: comments on https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/issues/559